common/config_manager.py

```diff
@@ -1,242 +1,245 @@
-
-    :param str attribute_name:  Setting to return
-    :param default:  Value to return if setting does not exist
-    :param connection:  Database connection, if None then a new connection will be created
-    :param cursor:  Database cursor, if None then a new cursor will be created
-    :param bool keep_connection_open:  Close connection after query?
-    :param bool raw:  True returns value as JSON serialized string; False returns JSON object
-    :return:  Setting value, or the provided fallback, or `None`.
-    """
-    if attribute_name in dir(ConfigManager):
-        # an explicitly defined attribute should always be called in favour
-        # of this passthrough
-        attribute = getattr(ConfigManager, attribute_name)
-        return attribute
-    else:
-        try:
-            if not connection or not cursor:
-                connection, cursor = quick_db_connect()
-
-            query = "SELECT value FROM settings WHERE name = %s"
-            cursor.execute(query, (attribute_name,))
-            row = cursor.fetchone()
-
-            if not keep_connection_open:
-                connection.close()
-
-            value = row.get("value", None) if row else None
-            if not raw and value is not None:
-                value = json.loads(value)
-        except (Exception, psycopg2.DatabaseError) as error:
-            raise ConfigException("Error getting setting {}: {}".format(attribute_name, repr(error)))
-
-        finally:
-            if connection is not None and not keep_connection_open:
-                connection.close()
-
-        if value is None:
-            # no value explicitly defined in the database...
-            if default is not None:
-                # if an explicit default was provided, return it
-                return default
-            elif attribute_name in config_definition and "default" in config_definition[attribute_name]:
-                # if a default is available from the config definition, return that
-                return config_definition[attribute_name]["default"]
-            else:
-                # use None as the last resort
+    db = None
+    dbconn = None
+    cache = {}
+
+    core_settings = {}
+    config_definition = {}
+    tag_context = []  # todo
+
+    def __init__(self, db=None):
+        # ensure core settings (including database config) are loaded
+        self.load_core_settings()
+        self.load_user_settings()
+
+        # establish database connection if none available
+        self.db = db
+
+    def with_db(self, db=None):
+        """
+        Initialise database
+
+        Not done on init, because something may need core settings before the
+        database can be initialised
+
+        :param db:  Database object. If None, initialise it using the core config
+        """
+        self.db = db if db else Database(logger=None, dbname=self.get("DB_NAME"), user=self.get("DB_USER"),
```

Comment:
I am wondering if there is a way to use 4CAT's `common.lib.logger`. It uses `config.get()` so we have an issue there _if_ this is called by one of those `get` usages. I see that Database now imports `logging` though.